### ISSUE TITLE
@Databases parameter does not work with databases that have a comma in the database name

### ISSUE THREAD

**Issue by johntrollope** [olahallengren/sql-server-maintenance-solution#314]
It is in fact possible to have a comma in a database name; I know this because one such database exists on my estate (some weird program that generates names based on user input, awful system, avoid it but I have to live with it).

To replicate:

CREATE DATABASE [a,test] ON 
( NAME = N'test', FILENAME = N'D:\test.mdf')
LOG ON 
( NAME = N'test_log', FILENAME = N'D:\test_log.ldf')

Hallengren's will pick this database up with USER_DATABASES but the following will fail:

EXECUTE dbo.IndexOptimize
@Databases = 'a,test',
@FragmentationLow = NULL,
@FragmentationMedium = NULL,
@FragmentationHigh = NULL,
@UpdateStatistics = 'ALL'

The error message is "The following databases in the @Databases parameter do not exist: [a], [test]." I also tried "[a,test]" but this also fails.

To search for an alternative delimiter, I did an extensive search on what characters are allowed in database names and crazy as it sounds, all these characters can be in the name of a database: !"£$%^&*()_+{}[@~'#<>?,./ including the space character plus all sorts of other non-ascii characters as long as the file names for the ldf and mdf are valid for DOS (and presumably Linux). The one and only one character which is not valid in a database name is the "]" (note that "[" is fine).

CREATE DATABASE [[] ON 
( NAME = N'test', FILENAME = N'D:\test2.mdf')
LOG ON 
( NAME = N'test_log', FILENAME = N'D:\test2_log.ldf')

This above works. But the below fails:

CREATE DATABASE []] ON 
( NAME = N'test', FILENAME = N'D:\test5.mdf')
LOG ON 
( NAME = N'test_log', FILENAME = N'D:\test5_log.ldf')

So an easy workaroud would be to use the "]" as a delimiter but that would just be bizarre. Thus the delimiter should remain the comma but the function to parse out database names needs to recognise that any commas possibly be part of the database name. To make this achievable, I think we need to be able to assume that any commas part of the database name will be within square brackets.

To summarise, "[a,test],db_a,db_b,[[another,awful,test],[,],[hello world]" should parse out to the following:

-         a,test
-         db_a
-         db_b
-         another,awful,test
-         , 
-         hello world

And if you want an extra challenge, cope with this use case: "aaa,[a weird [ database], [[],ouch"

-         aaa
-         a weird [ database
-         [
-         ouch
 
But let's be fair on ourselves and ignore that!

I am thinking that converting the string to xml and doing a two-phased parse would be the only way but this could become unwieldy code for such an unusual use case. But I am sure someone will have a good idea.

John

**Comment by johntrollope** [comment]
I actually think I have an approach:

•	Count all the [ and the ]
•	If there are more ], then bomb out politely
•	If equal number, then parse on ],[ 
•	In there are more [, then actually still parse on ],[

And for phase 1 of fix, do not allow mix of databases with [] and those with not. That is, allow "[a_db],[a,test]", but not "a_db,[a,test]

**Comment by srutzky** [comment]
Hello @johntrollope . This is certainly an understandable dilemma, but the analysis is incorrect.

> The one and only one character which is not valid in a database name is the "]" (note that "[" is fine).

That is not true:

1. A right square-bracket _is_ valid, it just needs to be escaped by doubling it (similar to using two single-quotes to escape a single quote within a string literal). Meaning: `[]]]` is a valid object name.
1. There are actually two characters which are invalid in an identifier name: `NCHAR(0x0000)` and `NCHAR(0xFFFF)` (or you can replace those hex values with their decimal equivalents of `0` and `65535`, respectively).

And, to prove that point, the following statement succeeds (at least in Windows, given that I let it use the DB name for the filename):

```sql
CREATE DATABASE []]];
```

If you are interested in the details (including the _complete_ list of rules for T-SQL identifiers), please see:

* [What’s in a Name?: Inside the Wacky World of T-SQL Identifiers](https://sqlquantumleap.com/2018/04/09/whats-in-a-name-inside-the-wacky-world-of-t-sql-identifiers/) (if not interested in the investigation, just scroll to bottom for the link to complete list)
* [The Uni-Code: The Search for the True List of Valid Characters for T-SQL Identifiers, Part 3 of 2 (Delimited Identifiers)](https://sqlquantumleap.com/2018/04/16/the-uni-code-the-search-for-the-true-list-of-valid-characters-for-t-sql-identifiers-part-3-of-2-delimited-identifiers/)

Of course, you still have your issue with a comma in a database name. For that, there are two appropriate options. None of the options include accepting delimiters as entity names in strings should not be delimited. Delimiting identifiers should _only_ be done when they are used in the regular T-SQL context. True, the string containing the database name in this scenario will be used in dynamic SQL and will thus be the T-SQL context upon execution, but that is for the code to handle by wrapping object names in the `QUOTENAME()` built-in function. A delimited list is not a string that can be used directly in a T-SQL context.

1. The quickest and easiest solution for you is to simply switch to using a different delimiter. I would suggest the "Unit Separator" character: `NCHAR(31)` or `NCHAR(0x001F)` (they are equivalent). Then, update both **dbo.IndexOptimize** and your code to use `NCHAR(0x001F)` instead of the comma as the delimiter.
1. The best long-term solution would be to add an optional parameter to **dbo.IndexOptimize** for `@Delimiter NVARCHAR(50) = N','`. This would default to a comma and thus be fully backwards compatible. But it would also allow someone to pass in any character (or short sequence of characters) to use instead of a comma. Most people would not specify that parameter at all, but you would pass in `@Delimiter = NCHAR(0x001F)`.

Take care,
Solomon...
https://SqlQuantumLift.com/
https://SqlQuantumLeap.com/
https://SQLsharp.com/

**Comment by olahallengren** [comment]
@johntrollope and @srutzky, thank you for the great input.

First I should say that comma in a database name, and having the requirement to specify that database in the ```@Databases``` parameter is a rare case. (I have been aware of this limitation in the code, since I developed it, but I don't recall anyone reporting it until now.)

That said, it should of course work with all characters.

One idea would be to move to table valued parameters, and get away from the string parsing. It would make this code simpler and more robust. It would also be easier to problematically pass databases to the stored procedure. The downside is that it would be a breaking change and also that it is not supported in SQL Server 2005.

**Comment by srutzky** [comment]
> First I should say that comma in a database name, and having the requirement to specify that database in the `@Databases` parameter is a rare case. ... That said, it should of course work with all characters.  
> 
> One idea would be to move to table valued parameters, and get away from the string parsing. It would make this code simpler and more robust. ... The downside is that it would be a breaking change and also that it is not supported in SQL Server 2005.

Hello @olahallengren . Yes, it was for both of those reasons &mdash; comma in database name is an edge-case _and_ TVPs do not work in SQL Server 2005 &mdash; that I did not suggest either TVPs or even XML (which would work in SQL Server 2005 and not require the extra object of the User-Defined Table Type). Merely making the delimiter be configurable via input parameter seemed like the least impacting and least complicated change, as well as being the least amount of effort required to make a change.

Of course, the true "least" complicated change is not making any change at all. If this is truly the only time this has been requested, then perhaps it's a customization that @johntrollope makes, which shouldn't be that bad, even needing to make that change again after updating to new versions, given how easy it would be (and even easier if the delimiter is still hard-coded and not set via parameter since it's an isolated, local customization &mdash; doing just option 1, noted in my comment above, because @johntrollope has no need to make it configurable as that code won't be distributed to others who might want it to be a comma).

**Comment by olahallengren** [comment]
Thanks again for the great input.

I am not planning to do any changes to this now.

**Comment by johntrollope** [comment]
Thanks both to Ola and Soloman. Indeed I did not know about how to make a database called "]". I also agree the best solution would be allow a user-defined delimiter as that would have neatly solved my problem. I solved my problem locally by getting agreement to change the db name and I will have a a "no comma allowed in db name" policy.

**Comment by olahallengren** [comment]
>Hello @olahallengren . Yes, it was for both of those reasons — comma in database name is an edge-case and TVPs do not work in SQL Server 2005 — that I did not suggest either TVPs or even XML (which would work in SQL Server 2005 and not require the extra object of the User-Defined Table Type).

I have decided that I will no longer support SQL Server 2005 in the main version. I released a version in the weekend, that is using some small features that are not available in SQL Server 2005.
https://ola.hallengren.com/versions.html

(A SQL Server 2005 - compatible version is available as a separate download.)

**Comment by olahallengren** [comment]
>I also agree the best solution would be allow a user-defined delimiter as that would have neatly solved my problem.

Yes, that would be a good solution. It should then work, not only for the Databases parameter, but for all the string - parameters, in all the stored procedures.

>I solved my problem locally by getting agreement to change the db name and I will have a a "no comma allowed in db name" policy.

It is good that you could solve it like that.

**Comment by olahallengren** [comment]
I have added a string delimiter parameter.

It is in the branch feature_StringDelimiter.
https://github.com/olahallengren/sql-server-maintenance-solution/tree/feature_StringDelimiter

Here is how to use it:

```
EXECUTE dbo.DatabaseBackup
@Databases = '[a,test]',
@Directory = 'C:\Backup',
@BackupType = 'FULL',
@Verify = 'Y',
@Compress = 'Y',
@CheckSum = 'Y',
@StringDelimiter = '|'
```

**Comment by olahallengren** [comment]
I have released a version with a string delimiter parameter.
https://ola.hallengren.com/versions.html
